### DISCUSSION TITLE
Official docker image uses debian:testing base ?

### DISCUSSION THREAD

**Discussion by deepakkamath01** [syslog-ng/syslog-ng#4168] [Q&A] [answered]
While browsing the Dockerfile for syslog-ng found the base image used is "Debian:testing", is this supposed to be that way with official images using an image with a  "testing" tag ?, just curious.

Code: https://github.com/syslog-ng/syslog-ng/blob/94be2d627381eaf764c25cf416d8270a43cd4be2/docker/Dockerfile#L1

**Comment by bazsi** [comment]
This is how we can iterate on the latest and greatest from the OS itself and functionality wise, syslog-ng is thoroughly tested by its automated test suite.

We can argue that syslog-ng's dependencies from the base OS is very thin, quoting the syslog-ng-core package's control file:

Depends: libc6 (>= 2.34), libcap2 (>= 1:2.10), libglib2.0-0 (>= 2.57.2), libivykis0 (>= 0.42), libjson-c5 (>= 0.15), libnet1 (>= 1.1.2.1), libpcre3, libssl3 (>= 3.0.0), libsystemd0 (>= 245), libwrap0 (>= 7.6-4~), util-linux (>= 2.12-10), lsb-base (>= 3.0-6)

From these: 
  * libc6, libpcre3, util-linux are all essential packages and very stable even in Debian testing.
  * libglib2.0-0 has 3579 reverse dependencies, well maintained and not just for syslog-ng
  * libssl3 has 822 reverse dependencies, thus pretty stable by the time it hits testing. Frequent updates are important from a security standpoint.
  * libivykis0 is supplied by us in the production syslog-ng image (e.g. we are not using the system supplied version)
  * libjson-c is not used in basic use-cases, although its use has increased continuously. libjson-c is a pretty well maintained library, yet to see a breakage that happened because of a newer version.
  * libnet1 is only used for spoof-source() support in the udp() destination, rarely used these days.
  * libsystemd0 is not used in a docker based environment
  * libwrap0 is not really used anymore, Fedora even killed this as a feature. (it is part of the tcpwrappers functionality, e.g. /etc/hosts.allow and similar access control functions)

Truth be told, the latest and greatest we need because of stuff like librdkafka or mqtt (both are separate syslog-ng packages, not installed by default).  But we also want to supply those functionalities in our production image, and stable often lags behind in some of these external dependencies.

Traditional syslog related use cases use very few things from the OS and those we consider stable enough in testing and we are yet to see any breakage that stemmed from this choice.

But let me know what you'd consider a risky dependency. We could consider an alternative image with a different base OS, but I am afraid it wouldn't provide all the features the current one is able to provide.

**Comment by bazsi** [comment, marked as answer]
This is how we can iterate on the latest and greatest and syslog-ng is
thoroughly tested by its automated test suite.

Truth be told, the latest and greatest we need because of stuff like
librdkafka. Traditional syslog related use cases use very few things from
the OS, e.g. the libc, which is a well maintained piece of debian testing.
Everything else (ivykis is at a version we supply along with syslog-ng)

We are yet to see any breakage that stemmed from this choice.

But let me know what you'd consider a risky dependency

**Reply by deepakkamath01** [reply]
Thankyou for the detailed response. I understand now why you chose that. But was just curious if the "debian:latest" is far behind in functionality than "debian:testing".
